tests/php/core-api/test_class.jetpack-core-api-module-endpoints.php

```diff
@@ -0,0 +1,198 @@
+
+	public static $current_user_can = true;
+
+	public $redefines = array();
+
+	public function setUp() {
+		require_once dirname( __FILE__ ) . '/../../../_inc/lib/class.core-rest-api-endpoints.php';
+
+		parent::setUp();
+
+		Jetpack::load_xml_rpc_client();
+
+		$this->redefines[] = redefine( 'Jetpack_IXR_Client::query', function() {
+			return true; // noop
+		} );
+
+		$this->redefines[] = redefine( 'Jetpack_IXR_Client::getResponse', function() {
+			return WP_Test_Jetpack_Core_Api_Module_Activate_Endpoint::$xmlrpc_response;
+		} );
+
+		$this->redefines[] = redefine( 'current_user_can', function( $permission ) {
+			if ( 'jetpack_manage_modules' === $permission ) {
+				return WP_Test_Jetpack_Core_Api_Module_Activate_Endpoint::$current_user_can;
+			}
+			return false;
+		} );
+
+		$this->redefines[] = redefine( 'Jetpack::state', function( $slug ) { } );
+
+		$this->redefines[] = redefine( 'Jetpack::is_active', function() {
+			return true;
+		} );
+	}
+
+	public function tearDown() {
+		foreach ( $this->redefines as $redefine ) {
+			restore( $redefine );
+		}
+		parent::tearDown();
+	}
+
+	/**
+	 * @author zinigor
+	 * @covers Jetpack_Core_Module_Activate_Endpoint
+	 * @requires PHP 5.2
+	 */
+	public function test_Jetpack_Core_API_Module_Activate_Endpoint_class_structure() {
+
+		$reflection = new ReflectionClass( 'Jetpack_Core_API_Module_Activate_Endpoint' );
+
+		$this->assertFalse( $reflection->isAbstract() );
+		$this->assertTrue( $reflection->isSubclassOf( 'Jetpack_Core_API_XMLRPC_Consumer_Endpoint' ) );
+		$this->assertTrue( $reflection->implementsInterface( 'Jetpack_Core_API_Writable' ) );
+	}
+
+	/**
+	 * @author zinigor
+	 * @covers Jetpack_Core_API_Module_Activate_Endpoint
+	 * @requires PHP 5.2
+	 */
+	public function test_register_routes() {
+		$routes = $this->server->get_routes();
+		$this->assertArrayHasKey( '/jetpack/v4/module/(?P<slug>[a-z\-]+)/activate', $routes );
+
+		$route = $routes['/jetpack/v4/module/(?P<slug>[a-z\-]+)/activate'][0];
+		$this->assertInstanceOf(
+			'Jetpack_Core_API_Module_Activate_Endpoint',
+			$route['callback'][0]
+		);
+		$this->assertInstanceOf(
+			'Jetpack_Core_API_Module_Activate_Endpoint',
+			$route['permission_callback'][0]
+		);
+	}
+
+	/**
+	 * @author zinigor
+	 * @covers Jetpack_Core_API_Module_Activate_Endpoint
+	 * @requires PHP 5.2
+	 * @dataProvider permission_provider
+	 */
+	public function test_activate_module( $can_write ) {
+		self::$xmlrpc_response = true;
+		self::$current_user_can = $can_write;
+
+		Jetpack::deactivate_module( 'carousel' );
+
+		$request = new WP_REST_Request( 'POST', '/jetpack/v4/module/carousel/activate' );
+		$response = $this->server->dispatch( $request );
+
+		if ( $can_write ) {
+			$this->assertEquals( 200, $response->status );
+
+			$this->assertTrue( Jetpack::is_module_active( 'carousel' ) );
+		} else {
+			$this->assertEquals( 403, $response->status );
+
+			$this->assertFalse( Jetpack::is_module_active( 'carousel' ) );
+		}
+	}
```

Comment:
It's common to only test one thing, whereas this tests if both the `can_write` and `! can_write` cases work. There are a few ways this can be done. One way is to create separate test functions while another is to create a data provider with input and expected output.